### PR TITLE
`Lectures`: Fix undefined course issue in lecture attachments

### PR DESCRIPTION
**Pull request by eceeeren**
**NOTE: Please only test on Non-Chromium Browsers! (Mozilla Firefox, Safari etc.)**

#### General

- [x] I tested **all** changes and their related features with **all** corresponding user types on a test server.
- [x] This is a small issue that I tested locally and was confirmed by another developer on a test server.
- [x] Language: I followed the [guidelines for inclusive, diversity-sensitive, and appreciative language](https://docs.artemis.cit.tum.de/dev/guidelines/language-guidelines/).
- [x] I chose a title conforming to the [naming conventions for pull requests](https://docs.artemis.cit.tum.de/dev/development-process/development-process.html#naming-conventions-for-github-pull-requests).

#### Client
- [x] **Important**: I implemented the changes with a very good performance, prevented too many (unnecessary) REST calls and made sure the UI is responsive, even with large data (e.g. using paging).
- [x] I **strictly** followed the [client coding and design guidelines](https://docs.artemis.cit.tum.de/dev/guidelines/client/).
- [x] I added multiple integration tests (Jest) related to the features (with a high test coverage), while following the [test guidelines](https://docs.artemis.cit.tum.de/dev/guidelines/client-tests/).
- [x] I documented the TypeScript code using JSDoc style.
- [x] I added multiple screenshots/screencasts of my UI changes.

### Motivation and Context

Lecture Attachments component is used in 2 ways:
1. In Lecture Creation Wizard
2. As the main view

On the main view of Attachments page, when we add a new attachment and direct to the View page, and when we come back to the Attachments page right after, lecture.course element becomes undefined. Therefore Course name in the title disappears and as View url has Course id in it, that becomes unusable as well.

This doesn't happen in wizard because every time we invoke it, a database call is made to fetch full lecture details.

**1. Course name before (undefined)**
Screenshot 2024-10-26 at 14 24 00

**2. View button before**
Screenshot 2024-10-26 at 14 24 13

### Description

Course element also added to the ActivatedRoute property as data and then set as lecture's course again.

### Steps for Testing

Prerequisites:
- 1 Instructor

1. Log in to Artemis
3. Navigate to Course Administration
4. Navigate to a Course's Lectures
5. Navigate to Attachments of any lecture
6. Add a new attachment
7. Go back to the Attachments page using the breadcrumbs or Cancel button
8. Course name should be displayed correctly near lecture name and View button should work

[](https://artemis-test1.artemis.cit.tum.de)
[](https://artemis-test2.artemis.cit.tum.de)
[](https://artemis-test3.artemis.cit.tum.de)
[](https://artemis-test4.artemis.cit.tum.de)
[](https://artemis-test5.artemis.cit.tum.de)
[](https://artemis-test6.artemis.cit.tum.de)
[](https://artemis-test9.artemis.cit.tum.de)

#### Code Review
- [ ] Code Review 1
- [ ] Code Review 2
#### Manual Tests
- [ ] Test 1
- [ ] Test 2

### Test Coverage

| Class/File | Line Coverage | Confirmation (assert/expect) |
|------------|--------------:|-----------------------------:|
| lecture-attachments.component.ts | 96% | ✅              |

### CODE REVIEW

**Review by HawKhiem — COMMENTED** [review]
I tested on TS2. After uploading the attachment, I could indeed view the attachment and the lecture name is displayed correctly. But after that when I click on cancel, the view button stops working again, with every click resulting in a bad request. I tried deleting the attachment and upload it again, but I couldn't


https://github.com/user-attachments/assets/94d47626-014b-4c56-bee2-d51c457f40d4

**Comment by eceeeren** [comment]
> I tested on TS2. After uploading the attachment, I could indeed view the attachment and the lecture name is displayed correctly. But after that when I click on cancel, the view button stops working again, with every click resulting in a bad request. I tried deleting the attachment and upload it again, but I couldn't
> 
>  Attachments.of.lecture._.Object.Orientation._.Practical.Course_.Interactive.Learning.SS24.-.Google.Chrome.2024-10-26.22-20-19.mp4

The PR is not ready to be reviewed yet

File: src/main/webapp/app/lecture/attachment.service.ts [hunk 1]
```diff
@@ -6,6 +6,7 @@ import { createRequestOption } from 'app/shared/util/request.util';
 import { Attachment } from 'app/entities/attachment.model';
 import { convertDateFromClient, convertDateFromServer } from 'app/utils/date.utils';
 import { objectToJsonBlob } from 'app/utils/blob-util';
+import _ from 'lodash';
```
Comment (krusche) [inline src/main/webapp/app/lecture/attachment.service.ts:9]:
```suggestion
import _ from 'lodash-es';
```

File: src/main/webapp/app/lecture/attachment.service.ts [hunk 2]
```diff
@@ -96,11 +97,10 @@ export class AttachmentService {
     }
 
     convertAttachmentDatesFromClient(attachment: Attachment): Attachment {
-        const copy: Attachment = Object.assign({}, attachment, {
+        return Object.assign({}, _.cloneDeep(attachment), {
```
Comment (krusche) [inline src/main/webapp/app/lecture/attachment.service.ts:100]:
```suggestion
        return Object.assign({}, cloneDeep(attachment), {
```

Comment (krusche) [inline src/main/webapp/app/lecture/attachment.service.ts:100]: Can you add a comment why this is necessary please?

**Comment by eceeeren** [comment]
> I tested on TS3, I am on a Mac using Chrome.
> 
> When I try to upload a new attachment in the course-attachment view I get the error-code 5 in my browser.

Can you try to remove your cookies? I had the same on Google Chrome until I removed my cookies but I thought I was the only one

**Comment by eceeeren** [comment]
> Tested on TS2. Every time I click on the "Save Attachment" button, when trying to create a new attachment, it leads to this page: image
> 
> Clicking the refresh button takes me back to the attachment page, without having saved the new attachment.

Can you try removing your cookies?

**Review by zagemello — COMMENTED** [review]
Managed to get it to work by refreshing again, all works as expected.

**Review by Cathy0123456789 — COMMENTED** [review]
Tested on TS5. Works after deleting the cookies.

**Review by flbrgit — COMMENTED** [review]
Tested on TS5, also had to delete cookies but afterwards everything worked as expected

**Review by sawys777 — APPROVED** [review]
Tested on TS1, everything worked accordingly to the test description, did not need to delete any cookies and the name was correctly displayed and the view button was functioning as well.

**Review by krusche — APPROVED** [review]

**Pull request merged**